salt/metalk8s/orchestrate/bootstrap/init.sls

```diff
@@ -1,23 +1,33 @@
 {# Because of the grain lookup below, the bootstrap minion *must* be available
    before invoking this SLS, otherwise rendering will fail #}
-{%- set bootstrap_control_plane_ip = salt.saltutil.cmd(
-        tgt=pillar.bootstrap_id,
-        fun='grains.get',
-        kwarg={
-            'key': 'metalk8s:control_plane_ip',
-        },
-    )[pillar.bootstrap_id]['ret']
-%}
-
-{%- set bootstrap_workload_plane_ip = salt.saltutil.cmd(
-        tgt=pillar.bootstrap_id,
-        fun='grains.get',
-        kwarg={
-            'key': 'metalk8s:workload_plane_ip',
-        },
-    )[pillar.bootstrap_id]['ret']
-%}
 
+{%- set max_try = 5 %}
+{%- set bootstrap_grains = {} %}
+{%- for _ in range(1, max_try) %}
+  {%- set res = salt.saltutil.cmd(
+          tgt=pillar.bootstrap_id,
+          fun='grains.get',
+          kwarg={
+              'key': 'metalk8s:control_plane_ip',
+          },
+      )
+  %}
+  {%- if pillar.bootstrap_id in res and 'ret' in res[pillar.bootstrap_id] %}
+    {%- do bootstrap_grains.update(
+        {'control_plane_ip': res[pillar.bootstrap_id]['ret']}
+    ) %}
+    {%- break %}
+  {%- endif %}
+{%- endfor %}
+{%- if 'control_plane_ip' not in bootstrap_grains %}
+Get metalk8s:control_plane_ip grain:
+  test.fail_without_changes:
+    - comment: >-
+        Unable to get metalk8s:control_plane_ip grain from {{ pillar.bootstrap_id }}
+        after {{ max_try }} tries.
+    - require_in:
+      - salt: Deploy CA role on bootstrap minion
```

Comment:
IMHO you should raise in jinja here instead of using a state, especially since I think if control plane ip not available it will raise in jinja when trying to render the pillar_data dict bellow

Reply:
I tried to raise with jinja first, but the traceback is really ugly and not helpful.

Reply:
Wouldn't it be better if templates never raised? Salt output is already not the easiest to parse, if we need to also extract messages from exception tracebacks... :sob:

Reply:
You would still need to put all the rest of this formula in an `{% else %}` block, to avoid generating another traceback few lines below.

Reply:
> Edit: ah no, since bootstrap_grains['control_plane_ip'] would simply return a jinja2.Undefined object

Really ? Didn't know.

Ok fair ! Maybe we want a `failhard` then. I think for that kind of check it make sense